src/util/memory.rs

```diff
@@ -15,6 +16,14 @@ const MMAP_FLAGS: libc::c_int = libc::MAP_ANON | libc::MAP_PRIVATE | libc::MAP_F
 // MAP_FIXED is used instead of MAP_FIXED_NOREPLACE (which is not available on macOS). We are at the risk of overwriting pre-existing mappings.
 const MMAP_FLAGS: libc::c_int = libc::MAP_ANON | libc::MAP_PRIVATE | libc::MAP_FIXED;
 
+/// This static variable controls whether we annotate mmapped memory region using `prctl`. It can be
+/// set via `Options::mmap_annotation` or the `MMTK_MMAP_ANNOTATION` environment variable.
+///
+/// FIXME: Since it is set via `Options`, it is in theory a decision per MMTk instance. However, we
+/// currently don't have a good design for multiple MMTk instances, so we use static variable for
+/// now.
+pub(crate) static MMAP_ANNOTATION: AtomicBool = AtomicBool::new(true);
```

Comment:
This looks bad. Though there is a `FIXME` above, we should avoid introducing this at first place.

Both `Plan` and `Space` have references to the options. You can let plan/space to check options and create a `MmapAnnotation`. If the runtime option is not enabled, we can use `Option<MmapAnnotation>` or `MmapAnnotation::Omit/None`.

Reply:
Yes.  That looks bad.

Although it is possible to let plans and spaces check options, and optionally pass `MmapAnnotation::Omit` or a `None` for `Option<MmapAnnotation>`, it is not ideal.

One important part of my intention is that *the annotation cannot be omitted*.  I'd like the function signatures to enforce that *an annotation must be provided for every call to mmap*.  If we give the caller the option to pass `None` or `Omit`, developers will do it if they think it is not worth annotating.  Instead, the decision of whether to actually annotate the mmap should be postponed until `mmap` is actually called and is successful, and should be controlled by the user via an option rather than at call sites.

And checking at every call site is repetitive and error prone.  Developers may forget to do the check or erroneously omit the annotation.

Ideally, functions related to memory mapping should be wrapped into a struct, maybe named `MmapSupport` or `CommonMapper`.  It will be created per `MMTK` instance, like `Mmapper` and `VMMap` (which are currently both global singletons), and referenced by plans, spaces, SFT, etc., whatever needs to use mmap.  The `mmap_annotation: bool` should be an instance variable of `CommonMapper`, and `map_fixed` should be an instance method of `MmapSupport`.  That may involve much refactoring, including moving `dzmmap`, `dzmmap_noreplace` and `mmap_noreserve` into `MmapSupport`.  For now, creating a global boolean variable seems to be the easiest solution.

In the short term, we may introduce `MmapSupport` as a global singleton, like `Mmapper` and `VMMap`.  In the long term, we can make them specific to the `MMTK` instance.

Reply:
> One important part of my intention is that the annotation cannot be omitted. I'd like the function signatures to enforce that an annotation must be provided for every call to mmap. 

We could enforce the users to supply `MmapAnnotation`, and turn it into `Option<MmapAnnotation>` before we send it to mmap functions. Plus the problem of having a global var outweighs the benefits of 'enforcing annotations'.

Alternatively, we can have an instance for `memory` which holds the boolean value from the option. I think they are both worth trying before adopting the current implementation. We have arg structs for creating plans and spaces. Creating a `Memory` instance and letting plan/space to hold it shouldn't be too hard. https://github.com/mmtk/mmtk-core/blob/8640ab89f43f33af99e61b9ec8ff2164ed2cfab7/src/plan/global.rs#L372

> It will be created per MMTK instance, like Mmapper and VMMap (which are currently both global singletons), and referenced by plans, spaces, SFT, etc., whatever needs to use mmap. 

I think it is still arguable whether `Mmapper` and `VMMap` should be global or per instance. We do have a hacky `vm_layout()` that should be per instance, and is global at the moment. But we should avoid introducing more globals like this. The more we have, the more difficult and the more resistance we will face when we try to support multiple instances.

Reply:
> Alternatively, we can have an instance for `memory` which holds the boolean value from the option. I think they are both worth trying before adopting the current implementation. We have arg structs for creating plans and spaces. Creating a `Memory` instance and letting plan/space to hold it shouldn't be too hard.
> 
> https://github.com/mmtk/mmtk-core/blob/8640ab89f43f33af99e61b9ec8ff2164ed2cfab7/src/plan/global.rs#L372

Yes.  This is my intended way to do it.  But this may involve too many changes and I think it is better to do the refactoring separately.

The introduced instance (let's call it `MmapSupport` for now) should contain `dzmap`, `dzmap_noreplace`, `mmap_noreserve` and `mmap_fixed`.  I grepped the code and found that the following places will be affected.

-   `MapState`: `MapState::transition_to_*` will call those functions directly, as if the `MapState` can transition its own state without the help of `Mmapper` or the `MmapSupport` instance.  This needs to be refactored so that `MapState` will become a state holder, and `Mmapper` implementations (`FragmentedMapper` and `ByteMapMmapper`) can call the mmap functions in `MmapSupport` to do the transition.
-   `LockFreeImmortalSpace`: It skipped the `Mmapper` and directly calls `dzmmap_noreplace`.  I think this is a bug because `Mmapper` should be aware of the mapping of spaces.
-   `RawMemoryFreeList`: It calls `dzmmap_noreplace` to map its memory.  Not sure if it should go through `Mmapper`, but probably it should because a `RawMemoryFreeList` is part of a space using it.

In the end, we may have all mmap invocations going through `Mmapper`, and an `Mmapper` holds a reference to `MmapSupport`.

> We could enforce the users to supply `MmapAnnotation`, and turn it into `Option<MmapAnnotation>` before we send it to mmap functions.

It is good to enforce the users to supply `MmapAnnotation`.  The difficulty is drawing a border between the "user" and the "mmap functions".  And another difficulty is that we lost access to `Options` before reaching "mmap functions".

Take the call chain `SideMetadataContext::map_metadata_internal` -> `try_mmap_contiguous_metadata_space` -> `MMAPPER.ensure_mapped` for example.  

-   In `SideMetadataContext::map_metadata_internal`, we have enough information (space name and side metadata name) to construct the annotation.  We construct `MmapAnnotation` here.  But since `SideMetadataContext` does not reference any`Space` or `Plan`, it cannot access `Options`.
-   `try_mmap_contiguous_metadata_space` is a top-level function.  It has access to neither space name nor `Option`.
-   `MMAPER.ensure_mapped` is a function that is more related to mmap.  It may be refactored to hold a boolean variable `MMAP_ANNOTATION`, or refer to `MmapSupport` I mentioned above.

So the availability of `Options` or the `MMAP_ANNOTATION` variable does not match the distinction between user and mmap.  I think it's easier to refactor the code and just let "mmap functions" decide whether to annotate.

> Plus the problem of having a global var outweighs the benefits of 'enforcing annotations'.

Having global var is bad, but I think it is worse to let the users check `Options::mmap_annotation` before constructing the `Option<MmapAnnotation>` because it is repetitive and error-prone.  Having a global variable is the easiest workaround before we refactor `Mmapper`, `MapState`, etc.

If you think having a global variable is unacceptible, I can remove `Options::mmap_annotation` and make the annotation controlled by a Cargo feature instead, or leave it always on.  I think it is OK to leave it always on because there will be strictly fewer `prctl` than `mmap`, and we don't call `mmap` very often, either.

Reply:
> If you think having a global variable is unacceptible, I can remove Options::mmap_annotation and make the annotation controlled by a Cargo feature instead, or leave it always on.

Right. We dont really need the option anyway. We can always annotate the mmaps.

The reason that I am against having a global var is that we know it's wrong and this kind of small issues will accumulate over time and eventually make it hard to get rid of them. We should try to avoid introducing them in first place.

Reply:
I removed the option, and added a Cargo feature `no_mmap_annotation` just in case anyone needs to disable it for any reason.

Reply:
This global var can be removed now. I am suprised that we did not get any warning from the compiler for this, this seems to be an unused variable.